stm32: clock_control: Add test suite for stm32

FRASTM opened the pull request:
Similarly to what was done for U5 in https://github.com/zephyrproject-rtos/zephyr/pull/40731 and H7 in https://github.com/zephyrproject-rtos/zephyr/pull/41391
add some tests dedicated to other stm32 clock_control driver configuration

- stm32g0 target board : nucleo_g071rb : with a **clear_clocks**  overlay to clear clocks then apply one config PLL 64MHz from HSI 16 MHz, PLL 64 MHz from HSE 8MHz.
  To set the PLL source to HSE, it is required to have the SB17 (MCO) closed to feed the stm32g071 HSE from the STLINK

- stm32g4 target board : nucleo_g474re : with a **clear_clocks**  overlay to clear clocks then apply one config PLL 64MHz from HSI 16 MHz, PLL 170 MHz from HSE 24MHz, 
  HSE 24MHz (no PLL) HSE is not bypassed. 

- stm32l1 and stm32l0 target board : nucleo_l152re and nucleo_l073rz : with a **clear_clocks**  overlay to clear clocks then apply one config PLL 32MHz from HSI 16 MHz, PLL 32MHz from HSE 8MHz (HSE is  bypassed),  HSE 8MHz (no PLL) HSE is  bypassed, HSI 16MHz, msi range 6 (also valid for other msi_range values).

- stm32wl55 target board : nucleo_wl55jc : with a **wl_clear_clocks**  overlay, config PLL 48MHz from HSI 16 MHz, PLL 48MHz from HSE 32MHz (HSE is  bypassed and HSE prescaler), HSE 32MHz (no PLL) (HSE is  bypassed), HSI 16MHz, msi range11 (also valid for other msi_range values).


- stm32wb55 target board : nucleo_wb55rg : with a **wb_clear_clocks**  overlay, config PLL 48MHz from HSI 16 MHz, PLL 64MHz from HSE 32MHz, PLL 32MHz from MSI 4MHz, HSE 32MHz (no PLL), HSI 16MHz, msi range11 (also valid for other msi_range values). Note that the hse prescaler is not avaible in the clock config.


- stm32l4 and stm32l5 target board : with a **clear_clocks**  overlay plus **clear_msi_clocks**  overlay to clear clocks then apply one config PLL 64MHz from HSI 16 MHz, PLL 48MHz from MSI 4 MHz, PLL  64MHz from HSE 8MHz (HSE is  bypassed),  HSE 8MHz (no PLL) HSE is  bypassed, HSI 16MHz, msi range 11 (also valid for other msi_range values).


- stm32f0/f1/f2/f3/f4/f7 target board : with a **clear_fx_clocks**  overlay overlay to clear clocks then apply one config PLL 64MHz from HSI 16 MHz or 8MHz, PLL  64MHz from HSE 8MHz (HSE is  bypassed),  HSI 8MHz or 16Mh, HSE 8MHz (no PLL) HSE is  bypassed.



Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
tests: drivers: stm32 clock control testing on stm32g0

target is stm32g0 with pll 64MHz from hsi clock config
target is stm32g0 with pll 64MHz from hse clock config
target is stm32g0 with hsi clock config (no pll)

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
tests: drivers: clock_control for stm32h7 config

fix comment when configuring the PLL with CSI clock source

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
$ west build -p auto -b nucleo_g071rb -- -DDTC_OVERLAY_FILE="boards/clear_clocks_f0_g0.overlay" -DDTC_OVERLAY_FILE="boards/pll_64_hsi_16.overlay" tests/drivers/clock_control/stm32_clock_configuration/stm32_common/

FRASTM committed:
dts: bindings: clock stm32g4 has a PLL P divider

The PLL div-p for the stm32g4x has more possible value than stm32l4
possible ra nge is from 2 to 31

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
tests: drivers: stm32 clock control testing on stm32g4

target is stm32g4 with pll 64MHz from hsi clock config
target is stm32g4 with pll 64MHz from hse clock config
target is stm32g4 with hsi clock config (no pll)

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
adding stm32g474re target
Keeping the the ahb-prescaler, apb1-prescaler, apb2-prescaler as they are set by each board DTS (not cleared by the clear_clocks.overlay as hey are not required for the test case).

FRASTM committed:
tests: drivers: stm32 clock control testing on stm32l1 and stm32l0

target is stm32l1/l0 with pll 32MHz from hsi clock config
target is stm32l1/l0 with pll 32MHz from hse clock config
target is stm32l1/l0 with hse clock config (no pll)
target is stm32l1/l0 with hsi clock config (no pll)
target is stm32l1/l0 with msi clock config (no pll)

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
adding stm32l073rz and nucleo_l152re targets
HSE is bypassed, the max PLL freq is 32MHz, the MSI range is possible from 1 to 6

FRASTM committed:
tests: drivers: stm32 clock control testing on stm32wl55

target is stm32wl55 with clearing clock config
target is stm32wl55 with pll 48MHz from hsi clock config
target is stm32wl55 with pll 48MHz from hse clock config
target is stm32wl55 with hse clock config (no pll)
target is stm32wl55 with msi clock config (no pll)

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
adding target nucleo_wl55jc with:
`$ west build -p auto -b nucleo_wl55jc -- -DDTC_OVERLAY_FILE="boards/wl_clear_clocks.overlay;boards/pll_48_hsi_16.overlay" tests/drivers/clock_control/stm32_clock_configuration/stm32_common/`

erwango commented on tests/drivers/clock_control/stm32_clock_configuration/stm32_common/testcase.yaml:
```diff
@@ -0,0 +1,13 @@
+common:
+    timeout: 5
+    platform_exclude: nucleo_h723zg b_u585i_iot02a
```
This should be updated/removed

erwango commented on tests/drivers/clock_control/stm32_clock_configuration/stm32_common/src/test_stm32_clock_configuration.c:
```diff
@@ -0,0 +1,90 @@
+#include <drivers/clock_control.h>
+#include <drivers/clock_control/stm32_clock_control.h>
+#include <logging/log.h>
+LOG_MODULE_REGISTER(test);
```
Please put this out of the #include section

erwango commented on tests/drivers/clock_control/stm32_clock_configuration/stm32_common/boards/msi_range11.overlay:
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2022 STMIcroelectronics
```
```suggestion
 * Copyright (c) 2022 STMicroelectronics
```
Apply everywhere

FRASTM commented:
adding target nucleo_wb55rg with:
$ west build -p auto -b nucleo_wb55rg -- -DDTC_OVERLAY_FILE="boards/wb_clear_clocks.overlay;boards/pll_48_hsi_16.overlay" tests/drivers/clock_control/stm32_clock_configuration/stm32_common/

FRASTM committed:
tests: drivers: stm32 clock control testing on stm32wb55

target is stm32wb55 with clearing clock config
target is stm32wb55 with pll 48MHz from hsi clock config
target is stm32wb55 with pll 48MHz from msi clock config
target is stm32wb55 with pll 64Hz from hse clock config
target is stm32wb55 with hse, msi, hsi clock config (no pll)

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
tests: drivers: stm32 clock control for stm32l0 and stm32l1

Fix build error for stm32 devices which have no function
to get the PLL ON bit from the RCC_CR register
Use the register access instead.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
tests: drivers: stm32 hse clock control testing on stm32g071 nucleo

Testing the HSE on the nucleo_stm32g071rb requires a hw fixture
on the hw board : MCO signal must given by the STLink to the mcu.
Put a hardware fixture to activate the hse clock with by-passed
only if the SB17 is closed on the HW.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
adding target nucleo_l4xxx with:
$ west build -p auto -b nucleo_l476rg -- -DDTC_OVERLAY_FILE="boards/clear_clocks.overlay;boards/clear_clocks_msi.overlay;boards/pll_64_hsi_16.overlay" tests/drivers/clock_control/stm32_clock_configuration/stm32_common/

FRASTM commented:
rebase on d2a56c5047f9e7f10b1b0df53d57917d3980bf72

FRASTM committed:
tests: drivers: stm32 clock control testing on stm32fx mcus

target is stm32fxx with clearing clock config
target is stm32fxx with pll from hsi clock config
target is stm32fxx with pll from hse clock config (with bypass)
target is stm32fxx with hse, hsi, clock config (no pll)

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
adding stm32f0/f1/f2/f3/f4/f7 targets:

`$ west build -p auto -b nucleo_f207zg -- -DDTC_OVERLAY_FILE="boards/clear_f2_f4_f7_clocks.overlay;boards/f2_f4_f7_pll_64_hse_8.overlay" tests/drivers/clock_control/stm32_clock_configuration/stm32_common/`

`$ west build -p auto -b nucleo_f334r8 -- -DDTC_OVERLAY_FILE="boards/clear_f0_f1_f3_clocks.overlay;boards/f0__f3_pll_32_hse_8.overlay" tests/drivers/clock_control/stm32_clock_configuration/stm32_common/`

`$ west build -p auto -b nucleo_f429zi -- -DDTC_OVERLAY_FILE="boards/clear_f2_f4_f7_clocks.overlay;boards/f2_f4_f7_pll_64_hse_8.overlay" tests/drivers/clock_control/stm32_clock_configuration/stm32_common/`

`$ west build -p auto -b nucleo_f767zi -- -DDTC_OVERLAY_FILE="boards/clear_f2_f4_f7_clocks.overlay;boards/f2_f4_f7_pll_64_hsi_16.overlay" tests/drivers/clock_control/stm32_clock_configuration/stm32_common/`

FRASTM commented:
split hse 8 configuration in 2:  with and without by-pass to run on the nucleo_wb55rg

FRASTM committed:
tests: drivers: stm32 clock control testing on stm32l4 / l5 mcus

target is stm32l4x/l5x with clearing clock config
target is stm32l4x/l5x with pll 64MHz from hsi clock config
target is stm32l4x/l5x with pll 48MHz from msi clock config
target is stm32l4x/l5x with pll 64MHz from hse clock config (with bypass)
target is stm32l4x/l5x with hse, hsi, msi clock config (no pll)

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on tests/drivers/clock_control/stm32_clock_configuration/stm32_common/testcase.yaml:
```diff
@@ -0,0 +1,120 @@
+    extra_args: DTC_OVERLAY_FILE="boards/clear_clocks.overlay;boards/pll_64_hsi_16.overlay"
+    platform_allow: nucleo_g071rb nucleo_g474re
+  drivers.stm32_clock_configuration.common.sysclksrc_msi_pll_64_hsi_16:
+    extra_args: DTC_OVERLAY_FILE="boards/clear_clocks.overlay;;boards/clear_clocks_msi.overlay;boards/pll_64_hsi_16.overlay"
```
```suggestion
    extra_args: DTC_OVERLAY_FILE="boards/clear_clocks.overlay;boards/clear_clocks_msi.overlay;boards/pll_64_hsi_16.overlay"
```

erwango commented:
@FRASTM The following test is failed on my side, can you have a check ?

`nucleo_wb55rg/tests/drivers/clock_control/stm32_clock_configuration/stm32_common/drivers.stm32_clock_configuration.common.sysclksrc_wb_48_msi`

erwango commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -35,13 +35,14 @@
 #define fn_mco2_prescaler(v) LL_RCC_MCO2_DIV_ ## v
 #define mco2_prescaler(v) fn_mco2_prescaler(v)
 
-/* Calculate MSI freq for the given range(at RUN range, not after standby) */
```
This commit was already merged in a different version. Let's remove it
